### PR TITLE
Tensor API reinvention

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
Caught the problem to extend existing tensor to work with preallocated memory, clean up it a bit to make clearer.
@apertovs @l-bat Have a look, please

File: tests/unit/test_tensor.cpp [hunk 1]
```diff
@@ -80,31 +85,50 @@ TEST(TensorTest, DynamicShape) {
...
-        ASSERT_FALSE(dst_tensor.empty());
-        for (size_t index = 0; index < 1 * 3 * 224 * 224; ++index) {
-            ASSERT_EQ(dst_tensor.data()[index], index);
-        }
-
-        deepworks::Tensor non_empty_tensor({1, 3, 16, 16});
-        ASSERT_ANY_THROW(src_tensor.copyTo(non_empty_tensor));
+    deepworks::Tensor src_tensor({1, 3, 224, 224});
+    EXPECT_FALSE(src_tensor.empty());
+    for (size_t index = 0; index < 1 * 3 * 224 * 224; ++index) {
```
Comment (l-bat) [inline tests/unit/test_tensor.cpp:90]: Why not tensor.total()?

Comment (TolyaTalamanov) [reply tests/unit/test_tensor.cpp]: Clean up tests a bit, but not dive deeper

**Review by l-bat — APPROVED** [review]
Thank you!

**Review by apertovs — APPROVED** [review]
Looks good to me

**Pull request merged**